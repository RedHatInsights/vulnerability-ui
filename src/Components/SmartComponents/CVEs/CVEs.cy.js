import React from 'react';
import { mount } from '@cypress/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import CVEs from './CVEs';
import ReducerRegistry from '../../../Utilities/ReducerRegistry';

import cvesResponse from '../../../../cypress/fixtures/cves.json';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';
import messages from '../../../../locales/en.json';
import { DEFAULT_PAGE_SIZE } from '../../../Helpers/constants';

beforeEach(() => {
    cy.intercept('GET', '**/api/vulnerability/v1/vulnerabilities/cves**', cvesResponse);

    mount(
        <IntlProvider locale="en" messages={messages}>
            <Provider store={ReducerRegistry.getStore()}>
                <Router>
                    <CVEs />
                </Router>
            </Provider>
        </IntlProvider>
    );
});

describe('CVE list table', () => {
    it.only('exists', () => {
        cy.get('[data-ouia-component-id="cves-table"]');

        cy.get('body').matchImage({
            maxDiffThreshold: 0.1
        });
    });

    it('has data', () => {
        cy.get('[data-ouia-component-id="cves-table"] tbody [data-label="CVE ID"]').should('have.length', DEFAULT_PAGE_SIZE);
    });
});

describe('CVE list table rows', () => {
    it('should have all rows collapsed by default', () => {
        cy.get('.pf-c-table__expandable-row:visible').should('have.length', 0);
    });

    it('should be expandable and collapsable', () => {
        cy.get('.pf-c-table__toggle').each((item) => cy.wrap(item).click());
        cy.get('.pf-c-table__expandable-row:visible').should('have.length', DEFAULT_PAGE_SIZE);

        cy.get('.pf-c-table__toggle').each((item) => cy.wrap(item).click());
        cy.get('.pf-c-table__expandable-row:visible').should('have.length', 0);
    });
});

describe('CVEs list table expand all button', () => {
    it('should expand and collapse all rows', () => {
        cy.get('[data-ouia-component-id=ExpandCollapseAll]').should('have.attr', 'aria-label', 'Expand all');
        cy.get('[data-ouia-component-id=ExpandCollapseAll]').click();
        cy.get('.pf-c-table__expandable-row:visible').should('have.length', DEFAULT_PAGE_SIZE);
        cy.get('[data-ouia-component-id=ExpandCollapseAll]').should('have.attr', 'aria-label', 'Collapse all');

        cy.get('[data-ouia-component-id=ExpandCollapseAll]').click();
        cy.get('.pf-c-table__expandable-row:visible').should('have.length', 0);
        cy.get('[data-ouia-component-id=ExpandCollapseAll]').should('have.attr', 'aria-label', 'Expand all');
    });
});
