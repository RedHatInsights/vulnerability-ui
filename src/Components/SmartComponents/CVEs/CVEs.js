import React, { useMemo, useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Stack, StackItem } from '@patternfly/react-core';
import { VULNERABILITIES_HEADER, CVES_ALLOWED_PARAMS } from '../../../Helpers/constants';
import { GenericError } from '../../PresentationalComponents/EmptyStates/EmptyStates';
import { createCveListByAccount } from '../../../Helpers/VulnerabilityHelper';
import { constructFilterParameters, updateRef, useCreateUrlParams } from '../../../Helpers/MiscHelper';
import BusinessRiskModal from '../Modals/BusinessRiskModal';
import StatusModal from '../Modals/CveStatusModal';
import SystemCveTable from './CVEsTable';
import SystemCveTableToolbar from './CVEsTableToolbar';
import DownloadReport from '../../../Helpers/DownloadReport';
import {
    changeCveListParameters,
    fetchCveListByAccount,
    selectCve,
    expandCve,
    clearCVEsStore
} from '../../../Store/Actions/Actions';

export const CVETableContext = React.createContext({});

export const CVEs = () => {
    const dispatch = useDispatch();
    const [CveStatusModal, setStatusModal] = useState(() => () => null);
    const [CveBusinessRiskModal, setBusinessRiskModal] = useState(() => () => null);
    const [urlParamsAllowed, setUrlParamsAllowed] = useState(false);
    const [isFirstLoad, setIsFirstLoad] = useState(true);

    const cveList = useSelector(
        ({ CVEsStore }) =>  CVEsStore.cveList
    );
    const parameters = useSelector(
        ({ CVEsStore }) => CVEsStore.parameters
    );
    const selectedCves = useSelector(
        ({ CVEsStore }) => CVEsStore.selectedCves
    );
    const openedCves = useSelector(
        ({ CVEsStore }) => CVEsStore.openedCves
    );
    const expandCveDescription = useSelector(
        ({ CVEsStore }) => CVEsStore.expandCveDescription
    );

    const cves = useMemo(() => createCveListByAccount(cveList), [cveList]);
    const [createUrlParams, urlParameters] = useCreateUrlParams(['show_irrelevant', ...CVES_ALLOWED_PARAMS]);

    const apply = (filterParams = {}) => {
        const params = constructFilterParameters(filterParams);
        dispatch(changeCveListParameters(params));
    };

    useEffect(() => {
        if (isFirstLoad) {
            apply(urlParameters);
            setIsFirstLoad(false);
        }
        else {
            dispatch(fetchCveListByAccount(parameters));
            urlParamsAllowed
                && createUrlParams({ ...parameters })
                || setUrlParamsAllowed(true);
        }
    }, [parameters]);

    useEffect(() => {
        return () => {
            dispatch(clearCVEsStore());
        };
    }, []);

    const processError = () => {
        return GenericError;
    };

    const handleCveSelect = (isSelected, cveNames) => {
        dispatch(selectCve(cveNames || []));
    };

    const downloadReport = format => {
        DownloadReport.exec(fetchCveListByAccount, parameters, format, 'cves');
    };

    const showBusinessRiskModal = cvesList => {
        const  { meta } = cves;
        setBusinessRiskModal(() => () =>
            (<BusinessRiskModal cves={cvesList} updateRef={() => updateRef(meta, apply)} />)
        );
    };

    const showStatusModal = cvesList => {
        const  { meta } = cves;
        setStatusModal(() => () =>
            (<StatusModal cves={cvesList} updateRef={() => updateRef(meta, apply)} />)
        );
    };

    const openCves = (cves) => {
        dispatch(expandCve(cves));
    };

    if (!cves.errors) {
        return (
            <CVETableContext.Provider
                value={{
                    cves,
                    params: parameters,
                    selectedCves,
                    openedCves,
                    expandCveDescription,
                    methods: {
                        apply,
                        downloadReport,
                        selectCves: handleCveSelect,
                        showBusinessRiskModal,
                        showStatusModal,
                        openCves
                    }
                }}
            >
                <CveBusinessRiskModal />
                <CveStatusModal />

                <Stack>
                    <StackItem>
                        <SystemCveTableToolbar showAllCheckbox />
                    </StackItem>
                    <StackItem>
                        <SystemCveTable header={VULNERABILITIES_HEADER} showAllCheckbox />
                    </StackItem>
                </Stack>
            </CVETableContext.Provider>
        );
    } else {
        return processError(cves.errors);
    }

};

export default CVEs;
