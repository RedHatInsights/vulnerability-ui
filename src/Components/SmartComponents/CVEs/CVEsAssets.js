
import { classNames, expandable, sortable, nowrap, wrappable } from '@patternfly/react-table';
import messages from '../../../Messages';
import { intl } from '../../../Utilities/IntlProvider';

export const CVES_DEFAULT_FILTERS = {
    affecting: 'true'
};

export const VULNERABILITIES_HEADER = [
    {
        title: intl.formatMessage(messages.cveId),
        key: 'synopsis',
        transforms: [sortable],
        columnTransforms: [wrappable],
        cellFormatters: [expandable],
        isShownByDefault: true,
        isUnhidable: true
    },
    {
        title: intl.formatMessage(messages.publishDate),
        key: 'public_date',
        transforms: [sortable, wrappable],
        columnTransforms: [classNames('hide-description')],
        isShownByDefault: true
    },
    {
        title: intl.formatMessage(messages.impact),
        key: 'impact',
        transforms: [sortable],
        columnTransforms: [nowrap],
        isShownByDefault: true
    },
    {
        title: intl.formatMessage(messages.cvssBaseScore),
        key: 'cvss_score',
        transforms: [sortable, wrappable],
        isShownByDefault: true
    },
    {
        title: intl.formatMessage(messages.systemsExposed),
        key: 'systems_affected',
        transforms: [sortable, wrappable],
        isShownByDefault: true
    },
    {
        title: intl.formatMessage(messages.businessRisk),
        key: 'business_risk',
        transforms: [sortable, wrappable],
        isShownByDefault: true
    },
    {
        title: intl.formatMessage(messages.status),
        key: 'status',
        transforms: [sortable, wrappable],
        isShownByDefault: true
    }
];
