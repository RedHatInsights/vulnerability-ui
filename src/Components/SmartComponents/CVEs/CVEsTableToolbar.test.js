import CVEsTableToolbar from './CVEsTableToolbar';
import { CVETableContext } from './CVEs';
import { mountWithIntl } from '../../../Helpers/MiscHelper';
import toJson from 'enzyme-to-json';
import {
    handleChangePage,
    handleSetPageSize, 
    removeFilters
} from '../../../Helpers/TableToolbarHelper';


jest.mock('../../../Helpers/TableToolbarHelper', () => (
    {
      ...(jest.requireActual('../../../Helpers/TableToolbarHelper')),
      handleChangePage: jest.fn(),
      handleSetPageSize: jest.fn(),
      removeFilters: jest.fn()
    }
));

const mockContext = {
    cves: {
        isLoading: false,
        meta: {
            test: 'test'
        },
        data: [
            {
                type: 'cve',
                id: 'CVE-2019-6454',
                exposed_systems_count: 1,
                business_risk_id: 0,
                business_risk_justification: null,
                status_id: 2,
                status_justification: "testhello",
                attributes: {
                    business_risk: "Not Defined",
                    business_risk_id: 0,
                    business_risk_text: null,
                    cve_status_id: 2,
                    cve_status_text: "testhello",
                    cvss2_score: null,
                    cvss3_score: "6.500",
                    description: "A new domain bypass",
                    impact: "Moderate",
                    public_date: "2020-06-09T17:00:00+00:00",
                    reporter: 1,
                    rule: null,
                    status: "On-Hold",
                    status_id: 2,
                    status_text: "testhello",
                    synopsis: "CVE-2020-0543",
                    affected_systems: 1,
                    rule: 'testRule'
                }
            }
        ]

    },
    params: {},
    selectedCves: [],
    selectedRowsRawData: [
        {
          attributes: {
            business_risk: 'Not Defined',
            business_risk_id: 0,
            business_risk_text: null,
            cvss2_score: null,
            cvss3_score: '8.600',
            description: 'There\'s a flaw in libxml2. An attacker who is able to submit a crafted file to be processed by an application linked with libxml2 could trigger a use-after-free. The greatest impact from this flaw is to confidentiality, integrity, and availability.',
            impact: 'Moderate',
            known_exploit: false,
            public_date: '2021-04-22T00:00:00+00:00',
            rules: [],
            status: 'Resolved',
            status_id: 2,
            status_text: null,
            synopsis: 'CVE-2019-6454',
            systems_affected: 497,
            systems_status_divergent: 497,
            status_justification: 'testhello',
            exposed_systems_count: 1,
            business_risk_justification: null 
          },
          id: 'CVE-2019-6454',
          type: 'cve'
        }
    ],
    expandedRows: [],
    isAllExpanded: false,
    methods: {
        apply: jest.fn(),
        downloadReport: jest.fn(),
        selectCves: jest.fn(),
        showBusinessRiskModal: jest.fn(),
        showStatusModal: jest.fn(),
        openCves: jest.fn()
    }
};

describe('CVEsTableToolbar', () => {
    it('Should render without errors', () => {
        const wrapper = mountWithIntl(
            <CVETableContext.Provider value={mockContext}>
                <CVEsTableToolbar/>
            </CVETableContext.Provider>
        ); 
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Should CVEsTableToolbarWithContext inherit default props', () => {
        let tempContext = { ...mockContext, methods: {} };
        const tempWrapper = mountWithIntl(
            <CVETableContext.Provider value={tempContext}>
                <CVEsTableToolbar/>
            </CVETableContext.Provider>
        );
        const { totalNumber, apply, downloadReport } = tempWrapper.find('CVEsTableToolbarWithContext').props();
        expect(totalNumber).toEqual(0);
        expect(apply()).toEqual(undefined);
        expect(downloadReport()).toEqual(undefined);
        tempWrapper.unmount();
        tempContext = {};
    });

    describe('actionsConfig:', () => {
        it('Should openCves be called with testCve if isAllExpanded = false', () => {
            const wrapper = mountWithIntl(
                <CVETableContext.Provider value={mockContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            ); 
            wrapper.find("Button").first().simulate("click");
            expect(mockContext.methods.openCves).toHaveBeenCalledWith(['CVE-2019-6454']);
        });
    
        it('Should openCves be called with empty array if isAllExpanded = true', () => {
            let tempContext = { ...mockContext, isAllExpanded: true };
            const tempWrapper = mountWithIntl(
                <CVETableContext.Provider value={tempContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            );
            tempWrapper.find("Button").first().simulate("click");
            expect(tempContext.methods.openCves).toHaveBeenCalledWith([]);
            tempWrapper.unmount();
            tempContext = {};
        });
        
        it('Should showStatusModal be called with item ID and status_id = 0', () => {
            const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
            const wrapper = mountWithIntl(
                <CVETableContext.Provider value={testContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            ); 
            const { actionsConfig: { actions } } = wrapper.find('PrimaryToolbar').props();
            actions[2].onClick();
            expect(mockContext.methods.showStatusModal).toHaveBeenCalledWith([{ id: 'CVE-2019-6454', "status_id": 2, "justification": null, "exposed_systems_count": 497 }]);
        });

        it('Should showBusinessRiskModal be called with item ID and business_risk_id = 0', () => {
            const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
            const wrapper = mountWithIntl(
                <CVETableContext.Provider value={testContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            ); 
            const { actionsConfig: { actions } } = wrapper.find('PrimaryToolbar').props();
            actions[1].onClick();
            expect(mockContext.methods.showBusinessRiskModal).toHaveBeenCalledWith([{ id: 'CVE-2019-6454', "business_risk_id": 0, "justification": null }]);
        });
    });
    
    describe('pagination:', () => {
        it('Should call handlePageChange on page change', () => {
              const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
              const wrapper = mountWithIntl(
                  <CVETableContext.Provider value={testContext}>
                      <CVEsTableToolbar/>
                  </CVETableContext.Provider>
              ); 
              const { pagination: { onSetPage } } = wrapper.find('PrimaryToolbar').props();
              onSetPage(null, 2);
              expect(handleChangePage).toHaveBeenCalledWith(null, 2, expect.any(Function));
        });

        it('Should call handleSetPageSize on page size change', () => {
            const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
            const wrapper = mountWithIntl(
                <CVETableContext.Provider value={testContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            ); 
            const { pagination: { onPerPageSelect } } = wrapper.find('PrimaryToolbar').props();
            onPerPageSelect(null, 2);
            expect(handleSetPageSize).toHaveBeenCalledWith(null, 2, expect.any(Function));
      });
    });

    describe('bulkSelect:', () => {
        it('Should handleOnCheckboxChange unselect all selected items', () => {
              const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
              const wrapper = mountWithIntl(
                  <CVETableContext.Provider value={testContext}>
                      <CVEsTableToolbar/>
                  </CVETableContext.Provider>
              ); 
              const { bulkSelect: { onSelect } } = wrapper.find('PrimaryToolbar').props();
              onSelect();
              expect(mockContext.methods.selectCves).toHaveBeenCalledWith(false, []);
        });
        it('Should handleOnCheckboxChange select all items', () => {
            const testContext = { ...mockContext, selectedCves: [] };
            const wrapper = mountWithIntl(
                <CVETableContext.Provider value={testContext}>
                    <CVEsTableToolbar/>
                </CVETableContext.Provider>
            ); 
            const { bulkSelect: { onSelect } } = wrapper.find('PrimaryToolbar').props();
            onSelect();
            expect(mockContext.methods.selectCves).toHaveBeenCalled();
      });
    });

    describe('activeFiltersConfig:', () => {
        it('Should call removeFilters with two parameters', () => {
              const testContext = { ...mockContext, selectedCves: ['CVE-2019-6454'] };
              const wrapper = mountWithIntl(
                  <CVETableContext.Provider value={testContext}>
                      <CVEsTableToolbar/>
                  </CVETableContext.Provider>
              ); 
              const { activeFiltersConfig: { onDelete } } = wrapper.find('PrimaryToolbar').props();
              onDelete(null, 'testitem');
              expect(removeFilters).toHaveBeenCalledWith('testitem', expect.any(Function), undefined, {"affecting": "true"});
        });
    });
});