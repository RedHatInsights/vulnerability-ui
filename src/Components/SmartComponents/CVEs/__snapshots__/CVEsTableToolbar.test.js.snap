// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVEsTableToolbar Should render snapshot in a state without errors 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-c-toolbar  ins-c-primary-toolbar"
    data-ouia-component-id="PrimaryToolbar"
    data-ouia-component-type="PF5/Toolbar"
    data-ouia-safe="true"
    id="ins-primary-data-toolbar"
  >
    <div
      class="pf-v5-c-toolbar__content"
    >
      <div
        class="pf-v5-c-toolbar__content-section"
      >
        <div
          class="pf-v5-c-toolbar__group pf-m-filter-group ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg"
        >
          <div
            class="pf-v5-c-toolbar__item"
          >
            <button
              aria-expanded="false"
              class="pf-v5-c-menu-toggle"
              data-ouia-component-id="BulkSelect"
              type="button"
            >
              <span
                class="pf-v5-c-menu-toggle__text"
              >
                <label
                  class="pf-v5-c-check"
                >
                  <input
                    aria-invalid="false"
                    aria-label="Select all"
                    checked=""
                    class="pf-v5-c-check__input"
                    data-ouia-component-id="BulkSelectCheckbox"
                    data-ouia-component-type="PF5/MenuToggleCheckbox"
                    data-ouia-safe="true"
                    name="toggle-checkbox"
                    type="checkbox"
                  />
                  <span
                    aria-hidden="true"
                    class="pf-v5-c-check__label"
                    id="toggle-checkbox"
                  >
                    1 selected
                  </span>
                </label>
              </span>
              <span
                class="pf-v5-c-menu-toggle__controls"
              >
                <span
                  class="pf-v5-c-menu-toggle__toggle-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 320 512"
                    width="1em"
                  >
                    <path
                      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                    />
                  </svg>
                </span>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-toolbar__item ins-c-primary-toolbar__filter"
          >
            <div
              class="pf-v5-l-split ins-c-conditional-filter"
            >
              <div
                class="pf-v5-l-split__item"
              >
                <button
                  aria-expanded="false"
                  aria-label="Conditional filter toggle"
                  class="pf-v5-c-menu-toggle ins-c-conditional-filter__group"
                  data-ouia-component-id="ConditionalFilterToggle"
                  type="button"
                >
                  <span
                    class="pf-v5-c-menu-toggle__text"
                  >
                    <span
                      class="pf-v5-c-icon pf-m-md"
                    >
                      <span
                        class="pf-v5-c-icon__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      class="ins-c-conditional-filter__value-selector"
                    >
                      CVE
                    </span>
                  </span>
                  <span
                    class="pf-v5-c-menu-toggle__controls"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__toggle-icon"
                    >
                      <svg
                        aria-hidden="true"
                        class="pf-v5-svg"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 320 512"
                        width="1em"
                      >
                        <path
                          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
              <div
                class="pf-v5-l-split__item pf-m-fill"
                data-ouia-component-id="ConditionalFilter"
              >
                <span
                  class="pf-v5-c-form-control pf-m-icon ins-c-conditional-filter"
                >
                  <input
                    aria-invalid="false"
                    aria-label="search-field"
                    data-ouia-component-id="ConditionalFilter"
                    data-ouia-component-type="PF5/TextInput"
                    data-ouia-safe="true"
                    id="search-cve"
                    placeholder="Search ID or description"
                    type="text"
                    value=""
                    widget-type="InsightsInput"
                  />
                  <span
                    class="pf-v5-c-form-control__utilities"
                  >
                    <span
                      class="pf-v5-c-form-control__icon"
                    >
                      <span
                        class="pf-v5-c-icon pf-m-md ins-c-search-icon"
                      >
                        <span
                          class="pf-v5-c-icon__content"
                        >
                          <svg
                            aria-hidden="true"
                            class="pf-v5-svg"
                            fill="currentColor"
                            height="1em"
                            role="img"
                            viewBox="0 0 512 512"
                            width="1em"
                          >
                            <path
                              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-v5-c-toolbar__item pf-m-spacer-sm"
        >
          <button
            aria-expanded="false"
            aria-label="Export"
            class="pf-v5-c-menu-toggle pf-m-plain"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 1024 1024"
              width="1em"
            >
              <path
                d="M975.8,636.9 L870.9,741.8 L457.9,328.6 C452.1,322.8 445.4,319.9 437.9,319.9 C430.4,319.9 423.7,322.8 417.9,328.6 L328.8,417.7 C323,423.5 320.1,430.2 320.1,437.7 C320.1,445.2 323,451.9 328.8,457.7 L742,870.7 L636.9,975.8 C610.5,1002.2 619.4,1024 656.8,1024 L956,1024 C1014.5,1024 1024,1013.7 1024,955.9 L1024,656.7 C1023.9,619.4 1002.2,610.5 975.8,636.9 Z M128,128 L896,128 L896,361.7 C896.007942,370.182681 899.389907,378.313788 905.4,384.3 L996.7,475.6 C1006.8,485.7 1024,478.5 1024,464.3 L1024,22.7 C1024,16.1 1021.9,10.7 1017.6,6.4 C1013.3,2.1 1007.9,0 1001.3,0 L22.7,0 C16.1,0 10.7,2.1 6.4,6.4 C2.1,10.7 0,16.1 0,22.7 L0,1001.3 C0,1007.9 2.1,1013.3 6.4,1017.6 C10.7,1021.9 16.1,1024 22.7,1024 L463.4,1024 C469.862884,1023.98894 475.684489,1020.0908 478.156232,1014.11925 C480.627976,1008.14769 479.264428,1001.27548 474.7,996.7 L383.4,905.4 C377.413788,899.389907 369.282681,896.007942 360.8,896 L128,896 L128,128 Z"
              />
            </svg>
          </button>
        </div>
        <div
          class="pf-v5-c-toolbar__item  ins-c-primary-toolbar__actions pf-m-spacer-sm"
        >
          <button
            aria-expanded="false"
            aria-label="kebab dropdown toggle"
            class="pf-v5-c-menu-toggle pf-m-plain"
            data-ouia-component-id="BulkActionsToggle"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-v5-svg"
              fill="currentColor"
              height="1em"
              role="img"
              viewBox="0 0 192 512"
              width="1em"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
              />
            </svg>
          </button>
        </div>
        <div
          class="pf-v5-c-toolbar__item ins-c-primary-toolbar__pagination"
        >
          <div
            class="pf-v5-c-pagination pf-m-compact"
            data-ouia-component-id="pagination-top"
            data-ouia-component-type="PF5/Pagination"
            data-ouia-safe="true"
            id="options-menu-top-pagination"
            style="--pf-v5-c-pagination__nav-page-select--c-form-control--width-chars: 2;"
          >
            <div
              class="pf-v5-c-pagination__total-items"
            >
              <b>
                0 - 0
              </b>
               of 
              <b>
                0
              </b>
               
            </div>
            <div>
              <button
                aria-expanded="false"
                aria-haspopup="listbox"
                class="pf-v5-c-menu-toggle pf-m-plain pf-m-text"
                id="options-menu-top-toggle"
                type="button"
              >
                <span
                  class="pf-v5-c-menu-toggle__text"
                >
                  <b>
                    0 - 0
                  </b>
                   of 
                  <b>
                    0
                  </b>
                   
                </span>
                <span
                  class="pf-v5-c-menu-toggle__controls"
                >
                  <span
                    class="pf-v5-c-menu-toggle__toggle-icon"
                  >
                    <svg
                      aria-hidden="true"
                      class="pf-v5-svg"
                      fill="currentColor"
                      height="1em"
                      role="img"
                      viewBox="0 0 320 512"
                      width="1em"
                    >
                      <path
                        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
            <nav
              aria-label="Pagination"
              class="pf-v5-c-pagination__nav"
            >
              <div
                class="pf-v5-c-pagination__nav-control"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to previous page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="previous"
                  data-ouia-component-id="OUIA-Generated-Button-plain-1"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="pf-v5-c-pagination__nav-control"
              >
                <button
                  aria-disabled="true"
                  aria-label="Go to next page"
                  class="pf-v5-c-button pf-m-plain pf-m-disabled"
                  data-action="next"
                  data-ouia-component-id="OUIA-Generated-Button-plain-2"
                  data-ouia-component-type="PF5/Button"
                  data-ouia-safe="true"
                  disabled=""
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v5-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 256 512"
                    width="1em"
                  >
                    <path
                      d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                    />
                  </svg>
                </button>
              </div>
            </nav>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-toolbar__content"
    >
      <div
        class="pf-v5-c-toolbar__content-section"
      >
        <div
          class="pf-v5-c-toolbar__item"
        >
          <span
            class="ins-c-chip-filters"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-link"
              data-ouia-component-id="ClearFilters"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              Reset filters
            </button>
          </span>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-toolbar__content pf-m-hidden"
      hidden=""
    >
      <div
        class="pf-v5-c-toolbar__group"
      />
    </div>
  </div>
</DocumentFragment>
`;
